gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java

```diff
@@ -530,10 +605,42 @@ public <T> T getOption(Key<T> key) {
     return options.getOption(key);
   }
 
+  /** {@inheritDoc} */
+  @Override
+  public void validateUniverseDomain() {
+    EndpointContext endpointContext = getEndpointContext();
+    Credentials credentials = getCredentials();
+    try {
+      if (!endpointContext.hasValidUniverseDomain(credentials)) {
```

Comment:
I split the logic into two methods since I had a few issues
1. gRPC and HttpJson have slightly different behaviors with handling Auth Exception retries. Putting them in the EndpointContext would have them share the same behavior. 
2. ApiExceptionFactory requires a StatusCode which is in each transport-specific module (https://github.com/googleapis/sdk-platform-java/blob/ed16261ac8c90fe588026e30d62d948e7ca1af13/gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcStatusCode.java#L40). It's inaccessible from the transport-neutral module.

Reply:
1. I think the different behavior is in the catch block, where we want to rethrow a retryable exception, but in the try block, it should be the same?
2. Yeah it's unfortunate that `StatusCode` is an interface, we can get around it by creating an inner class that overrides `getCode()` and returns `PERMISSION_DENIED`.

Reply:
> I think the different behavior is in the catch block, where we want to rethrow a retryable exception, but in the try block, it should be the same?

I might be misunderstanding what you mean. We check the retryable exception to see if it should be retried. The Retryable is bubbled up from the `endpointContext.hasValidUniverseDomain(...)` call in the try block. What do you mean by they should be the same?